esp/services/esdl_svc_engine/esdl_binding.cpp

```diff
@@ -2492,10 +2539,90 @@ int EsdlBindingImpl::HandleSoapRequest(CHttpRequest* request,
     return 0;
 }
 
+class CNSVariableSubstitutionHelper : public CInterfaceOf<IVariableSubstitutionHelper>
+{
+    IEspContext&      m_context;
+    IEsdlDefinition&  m_esdl;
+    StringBuffer      m_service;
+    StringBuffer      m_method;
+public:
+    CNSVariableSubstitutionHelper(IEspContext& context, IEsdlDefinition& esdl, const char* service, const char* method)
+        : m_context(context)
+        , m_esdl(esdl)
+        , m_service(service)
+        , m_method(method)
+    {
+        m_service.toLowerCase();
+        m_method.toLowerCase();
+    }
+    virtual bool findVariable(const char* name, StringBuffer &value) override
+    {
+        StringBuffer tmp;
+        bool         required = true;
+        if (!isEmptyString(name))
+        {
+            if (streq(name, "service"))
+            {   // service name, lowercase
+                tmp.append(m_service);
+            }
+            else if (streq(name, "esdl-service"))
+            {   // service name, possible mixed case
+                IEsdlDefService* esdlService = m_esdl.queryService(m_service);
+                if (esdlService)
+                    tmp.append(esdlService->queryName());
+            }
+            else if (streq(name, "method"))
+            {   // method name, lowercase
+                if (!m_method.isEmpty())
+                    tmp.append(m_method);
+            }
+            else if (streq(name, "esdl-method"))
+            {   // methodname, possible mixed case
+                IEsdlDefService* esdlService = m_esdl.queryService(m_service);
+                IEsdlDefMethod* esdlMethod = (esdlService ? esdlService->queryMethodByName(m_method) : nullptr);
+                if (esdlMethod)
+                    tmp.append(esdlMethod->queryName());
+            }
+            else if (streq(name, "optionals"))
+            {   // optional esdl optionals (e.g., "internal"), lowercase
+                IProperties* params = m_context.queryRequestParameters();
+                Owned<IPropertyIterator> esdlOptionals = m_esdl.queryOptionals()->getIterator();
+                ForEach(*esdlOptionals)
+                {
+                    const char* key = esdlOptionals->getPropKey();
+                    if (params->hasProp(key))
+                    {
+                        if (!tmp.isEmpty())
+                            tmp.append(',');
```

Comment:
These assumptions about how optional values are formatted in the namespace show why a scripted version of namespace configuration would potentially be valuable.

Reply:
Just a comment, we shouldn't do anything like this now, but technically I think you could include parameters in the variable name with the current replace variables implementation.